api/v1alpha1/nnfdatamovementprofile_webhook.go

```diff
@@ -0,0 +1,120 @@
+)
+
+// log is for logging in this package.
+var nnfdatamovementprofilelog = logf.Log.WithName("nnfdatamovementprofile-resource")
+
+// SetupWebhookWithManager will setup the manager to manage the webhooks
+func (r *NnfDataMovementProfile) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+// +kubebuilder:webhook:path=/validate-nnf-cray-hpe-com-v1alpha1-nnfdatamovementprofile,mutating=false,failurePolicy=fail,sideEffects=None,groups=nnf.cray.hpe.com,resources=nnfdatamovementprofiles,verbs=create;update,versions=v1alpha1,name=vnnfdatamovementprofile.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &NnfDataMovementProfile{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *NnfDataMovementProfile) ValidateCreate() (admission.Warnings, error) {
+	nnfdatamovementprofilelog.Info("validate create", "name", r.Name)
+
+	// If it's not pinned, then it's being made available for users to select
+	// and it must be in the correct namespace.
+	profileNamespace := os.Getenv("NNF_DM_PROFILE_NAMESPACE")
+	if !r.Data.Pinned && r.GetNamespace() != profileNamespace {
+		err := fmt.Errorf("incorrect namespace for profile that is intended to be selected by users; the namespace should be '%s'", profileNamespace)
+		nnfdatamovementprofilelog.Error(err, "invalid")
+		return nil, err
+	}
+
+	if err := r.validateContent(); err != nil {
+		nnfdatamovementprofilelog.Error(err, "invalid NnfDataMovementProfile resource")
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *NnfDataMovementProfile) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
+	nnfdatamovementprofilelog.Info("validate update", "name", r.Name)
+
+	obj := old.(*NnfDataMovementProfile)
+	if obj.Data.Pinned != r.Data.Pinned {
+		err := fmt.Errorf("the pinned flag is immutable")
+		nnfdatamovementprofilelog.Error(err, "invalid")
+		return nil, err
+	}
+	if obj.Data.Pinned {
+		// Allow metadata to be updated, for things like finalizers,
+		// ownerReferences, and labels, but do not allow Data to be
+		// updated.
+		if !reflect.DeepEqual(r.Data, obj.Data) {
+			msg := "update on pinned resource not allowed"
+			err := fmt.Errorf(msg)
+			nnfdatamovementprofilelog.Error(err, "invalid")
+			return nil, err
+		}
+	}
+
+	if err := r.validateContent(); err != nil {
+		nnfdatamovementprofilelog.Error(err, "invalid NnfDataMovementProfile resource")
+		return nil, err
+	}
+	return nil, nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *NnfDataMovementProfile) ValidateDelete() (admission.Warnings, error) {
+	nnfdatamovementprofilelog.Info("validate delete", "name", r.Name)
```

Comment:
The "delete" verb isn't listed in the kubebuilder line, so this won't get run, right? Putting the log message in there is a little misleading.